aptos-move/framework/aptos-stdlib/sources/cryptography/algebra.move

```diff
@@ -42,22 +222,60 @@ module aptos_std::algebra {
         serialize_internal<S, F>(element.handle)
     }
 
-    //
-    // (Public functions end here.)
-    // Native functions begin.
-    //
+    /// Get the order of structure `S`, a big integer little-endian encoded as a byte array.
+    public fun order<S>(): vector<u8> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        order_internal<S>()
+    }
+
+    /// Cast an element of a structure `S` to a parent structure `L`.
+    public fun upcast<S,L>(element: &Element<S>): Element<L> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        Element<L> {
+            handle: upcast_internal<S,L>(element.handle)
+        }
+    }
 
-    native fun deserialize_internal<G, F>(bytes: &vector<u8>): (bool, u64);
-    native fun add_internal<F>(handle_1: u64, handle_2: u64): u64;
-    native fun serialize_internal<G, F>(handle: u64): vector<u8>;
+    /// Try casting an element `x` of a structure `L` to a sub-structure `S`.
+    /// Return none if `x` is not a member of `S`.
+    ///
+    /// NOTE: Membership check in `S` is performed inside, which can be expensive, depending on the structures `L` and `S`.
+    public fun downcast<L,S>(element_x: &Element<L>): Option<Element<S>> {
+        abort_unless_cryptography_algebra_natives_enabled();
+        let (succ, new_handle) = downcast_internal<L,S>(element_x.handle);
+        if (succ) {
+            some(Element<S> { handle: new_handle })
+        } else {
+            none()
+        }
+    }
+
+    /// Hash an arbitrary-length byte array `msg` into structure `S` using the given hash suite `H`.
+    /// A unique domain separation tag `dst` is required
+    /// for each independent collision-resistent mapping involved in the protocol built atop.
```

Comment:
Add a note here:

/// Some hashing methods do not accept a `dst` and will abort if a non-empty one is provided.

Reply:
Let's have a `hash_to_no_dst()` operation for that in the future.

Reply:
Yes, we will. But can you still add the comment above?

Reply:
Fixed.

Reply:
The warning/comment above is still missing though.

Reply:
double checked, it's [here](https://github.com/aptos-labs/aptos-core/pull/6550/files#diff-fa0448de48c5f7decf5d70fa37a7227cd7754236f3759257b714ae17de923f57R256).